### ISSUE TITLE
Unclear dictionary loading status on Kyocera KYF31

### ISSUE THREAD

**Issue by sspanak** [sspanak/tt9#117]
It looks like the progress bar is not showing up on Kyocera KYF31. Moreover, there may be a general issue with reloading the dictionary multiple times.

Original comment by puffertron from [another issue](https://github.com/sspanak/tt9/issues/104):

>I am having another problem though, with loading the English dictionary. (Since the last beta, I didn't test for very long and I figured i was cancelling the loading accidentally by letting my phone fall asleep or exiting the menu screen)

>Loading a dictionary takes a while, but I can't tell if it's done or not after trying to load it. Prediction mode doesn't show any suggestions with real words, it just shows groups of letters based off the keys I've pressed. I think prediction is working because when I add a word manually it shows up in the prediction bar.

>Update: while writing this my phone has been trying to load the dictionary for more than 20 minutes now, I think there may be a bug here too.

**Comment by sspanak** [comment]
Github doesn't allow me to ping **puffertron** here. Anyway, I need extra information in order to figure this issue out.

I think there may be two separate problems, so I need to know the following:
1. Have you tried using "clear dictionary", then "load dictionary" or you are loading without clearing? Loading should not take that long. With an empty database, English loads in about 10 seconds on my phone.
2. Does your phone support notifications like regular Android phones? If yes, do you see these notifications:
![IMG_20221123_163107](https://user-images.githubusercontent.com/27846105/203575374-21dffcfa-7c2e-49be-b550-baeaae5b9e4f.jpg)
![IMG_20221123_162938](https://user-images.githubusercontent.com/27846105/203575353-2d7cbbab-ca23-4275-8eae-01d1ccff26d7.jpg)

**Comment by sspanak** [comment]
Loading runs in a separate thread, so it shouldn't be affected by the notifications, but I can check what happens if they throw an error.

Anyway, my feeling is progress notifications do not work on your phone or Android version. I don't think I can fix them without having the phone in my hands, so as an alternative, I am thinking of displaying the loading status on the Settings screen. But before attempting anything, I need to know the following:
1. When you attempt to load a dictionary, does "load dictionary" change to "cancel loading"?
2. If yes, what happens if you go to another application, then back to the TT9 settings? Do you still see "cancel loading" (provided that loading is still in progress)?
3. Does the [latest stable v10.0](https://github.com/sspanak/tt9/releases/tag/v10.0-new-candidate-view%2Bword-loader-in-notifications) behave any differently?

**Comment by puffertron** [comment]
> When you attempt to load a dictionary, does "load dictionary" change to "cancel loading"?

The text does not change to cancel loading when I hit the load button.

> If yes, what happens if you go to another application, then back to the TT9 settings? Do you still see "cancel loading" (provided that loading is still in progress)?

When I leave to another app and come back, the text does change to cancel loading, but if the loading completes while I wait on the page, the text will not change back to "load dictionaries" unless I exit and come back to settings.

In the stable version, dictionaries load with no problem.

**Comment by sspanak** [comment]
>In the stable version, dictionaries load with no problem.

Alright. And do you see the progress notification there? This is also important, but I forgot about it.

**Comment by sspanak** [comment]
I've run some tests on my old Android 4.4.2 tablet and fixed several bugs, that, strangely, were not occurring on Android 11.

Hopefully, the notifications will work on the Kyocera now, provided that it supports them. If not, dictionary loading progress is also available below to the button. The problems with going to another app, then back to TT9 are resolved too (at least on my devices).

@puffertron, if you have the time, [beta-10.2](https://github.com/sspanak/tt9/releases/tag/beta-10.2) is available now.

**Comment by puffertron** [comment]
Testing on the stable 10.0 build, dictionary loading seems to have a functioning loading notification. The dictionary loads and prediction works in this build.

In the newest 10.2 build the notification is not visible, and the dictionary sends the "dictionary loaded" toast message but does not actually load the dictionary.

A temporary solution I've found is that if I load the dictionary with the stable version and use 'adb install -r' to replace install the APK with the beta version, the dictionary stays loaded, which is nice.

**Comment by sspanak** [comment]
I see. Toast messages were not displaying errors, but now I made them to do so. Here is the new: [beta-10.3](https://github.com/sspanak/tt9/releases/tag/beta-10.3).

I could potentially enable some more error logging, but for now I am not willing to do it outside of the development environment. I'll do it only if the problem is still unclear.

> A temporary solution I've found is that if I load the dictionary with the stable version and use 'adb install -r' to replace install the APK with the beta version, the dictionary stays loaded, which is nice.

Yep, you could also install v10.2 over v10.0 the regular way and Android will preserve the local storage and the database. It is how it works, there is no magic. Using `abd` is just an alternative way to do it. :slightly_smiling_face:

**Comment by puffertron** [comment]
I tested beta 10.3 and I never got any error messages, the dictionaries still don't load either. 

In version 10.3, there is no loading notification and the button only switches to 'cancel loading' when I leave and re-enter the settings page

Also, a weird thing I've noticed is that when I hit the clear dictionaries button(once) I get 2 toast messages in quick sucession that say dictionary cleared.

**Comment by sspanak** [comment]
> Also, a weird thing I've noticed is that when I hit the clear dictionaries button(once) I get 2 toast messages in quick sucession that say dictionary cleared.

That's the key. This means the buttons are getting two click events. This causes the "load" button to start the dictionary load, but then immediately cancel it. I am unsure why this happens on your phone, but I've added a protection against fast clicking. I could be useful in other cases too.

I think I got it this time! Many thanks and you are welcome to try out [beta-10.4](https://github.com/sspanak/tt9/releases/tag/beta-10.4)

**Comment by sspanak** [comment]
No chance at all. I am using Linux and I never install random programs on my computer. It must be a false positive. This occasionally happens with large or compressed binaries. I've come across this problem some time ago, when I developed a tool in Golang. Searching the web revealed it is not an uncommon issue.

Anyway, I've built and attached a new `.apk` to the same release (and changed the title). I changed a couple of strings to make Android Studio generate a different file. Other than that, it is identical to the original 10.4. Please, go ahead and try it out.

I am sorry to bother you so much, but I am unable to test and verify this bug any other way. Thank you again!

**Comment by puffertron** [comment]
Yeah, i figured, just thought i should let you know. Thanks for rebuilding it!

Turns out the double click was the issue! The dictionary is loading fine and the notification is displaying. Sorry for all the runaround on this bug, i should have noticed the double toast message earlier.

> I am sorry to bother you so much, but I am unable to test and verify this bug any other way. Thank you again!

Don't worry about it! I'm happy to test the app as I'm sure I'll be using it regularly now that everything works on my phone :)

**Comment by sspanak** [comment]
Excellent! Now let's get it released... at last... :rocket:
